### ISSUE TITLE
stax convention for batch / single-instance?

### ISSUE THREAD

**Issue by KristianHolsheimer** [google/jax#2243]
Just a quick question. I'm writing some extensions for stax and I'm wondering if you have a convention regarding single-instance vs batch input.

For instance, in pytorch and keras all input is expected to be a batch of inputs, with the first axis being the batch axis by convention.

Looking at stax.Flatten, it does look like a similar convention is used in stax. Other transformations are compatible with both single-instance and batch inputs.

The reason I'm asking is that I found myself writing conditionals to deal with both, like
```python
if input.ndim >= 2:
  # something something vmap
else:
  # single-instance behavior
```
I prefer not to have to do this and assume that all input is batched.

Is the stax convention that all input is batched?

Thanks!

PS I love the jax project, keep up the good work!!

**Comment by shoyer** [comment]
Yes, stax is written to act on a batch of examples at once.

(There has been some interest in writing an “unbatched” neural net library in JAX and relying upon vmap for vectorization, but there are a few weird neural net layers like batch norm that act across a batch, and it isn’t clear yet that we can always match the performance of “pre-batched” neural net layers.)

**Comment by shoyer** [comment]
I’ll mark this as closed since the original question is answered, but feel free to follow-up

**Comment by KristianHolsheimer** [comment]
Thanks @shoyer !!
